File: bench/profile-assignment.js
"use strict";

const fs = require("node:fs");
const path = require("node:path");
const vm = require("node:vm");
const {performance} = require("node:perf_hooks");

const root = path.resolve(__dirname, "..");
const engineModuleFiles = [
  "state.js", "memo.js", "metrics.js", "topology.js", "board.js",
  "heuristic.js", "deadlock.js", "analysis.js", "push-generation.js",
];
const source = engineModuleFiles
  .map(file => fs.readFileSync(path.join(root, "src", file), "utf8"))
  .join("\n");
const context = {console, performance};
vm.runInNewContext(source, context, {filename: "solver-engine.js"});

function matrix(size, offset) {
  return Array.from({length: size}, (_, row) =>
    Array.from({length: size}, (_, column) =>
      ((row + 3) * (column + 5) + offset * 7) % 31));
}

function median(values) {
  return [...values].sort((left, right) => left - right)[Math.floor(values.length / 2)];
}

function time(operation, iterations) {
  const started = performance.now();
  for (let index = 0; index < iterations; index++) operation(index);
  return performance.now() - started;
}

const results = [];
for (let size = 2; size <= 10; size++) {
  const base = matrix(size, 0);
  const previous = context.minimumAssignment(base);
  const iterations = Math.max(500, Math.floor(12000 / size));
  const fullSamples = [], repairSamples = [];
  for (let repeat = 0; repeat < 7; repeat++) {
    fullSamples.push(time(index => {
      const changed = base.map(row => [...row]);
      changed[index % size] = matrix(size, index + repeat)[index % size];
      context.minimumAssignment(changed);
    }, iterations));
    repairSamples.push(time(index => {
      const changed = base.map(row => [...row]);
      const changedRow = index % size;
      changed[changedRow] = matrix(size, index + repeat)[changedRow];
      context.repairMinimumAssignment(previous, changed, changedRow);
    }, iterations));
  }
  const fullMs = median(fullSamples);
  const repairMs = median(repairSamples);
  results.push({
    size,
    iterations,
    fullMs: Math.round(fullMs * 1000) / 1000,
    repairMs: Math.round(repairMs * 1000) / 1000,
    ratio: Math.round((repairMs / fullMs) * 1000) / 1000,
  });
}
process.stdout.write(`${JSON.stringify({runtime: "javascript", results}, null, 2)}\n`);

File: bench/profile_assignment.py
"""Profile full Hungarian assignment against safe one-row repair."""

from __future__ import annotations

import json
import sys
import statistics
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from searches.Sokomind import _minimum_assignment, _repair_minimum_assignment


def matrix(size: int, offset: int) -> list[list[float]]:
    return [
        [float(((row + 3) * (column + 5) + offset * 7) % 31) for column in range(size)]
        for row in range(size)
    ]


def timed(operation, iterations: int) -> float:
    started = time.perf_counter()
    for index in range(iterations):
        operation(index)
    return (time.perf_counter() - started) * 1000


def main() -> None:
    results = []
    for size in range(2, 11):
        base = matrix(size, 0)
        previous = _minimum_assignment(base)
        iterations = max(500, 12_000 // size)
        full_samples = []
        repair_samples = []
        for repeat in range(7):

            def full(index: int) -> None:
                changed = [row.copy() for row in base]
                changed[index % size] = matrix(size, index + repeat)[index % size]
                _minimum_assignment(changed)

            def repair(index: int) -> None:
                changed = [row.copy() for row in base]
                changed_row = index % size
                changed[changed_row] = matrix(size, index + repeat)[changed_row]
                _repair_minimum_assignment(previous, changed, changed_row)

            full_samples.append(timed(full, iterations))
            repair_samples.append(timed(repair, iterations))
        full_ms = statistics.median(full_samples)
        repair_ms = statistics.median(repair_samples)
        results.append(
            {
                "size": size,
                "iterations": iterations,
                "fullMs": round(full_ms, 3),
                "repairMs": round(repair_ms, 3),
                "ratio": round(repair_ms / full_ms, 3),
            }
        )
    print(json.dumps({"runtime": "python", "results": results}, indent=2))


if __name__ == "__main__":
    main()
